modules/js-and-css/speculation-rules/class-plsr-url-pattern-prefixer.php

```diff
@@ -0,0 +1,91 @@
+<?php
+/**
+ * Class 'PLSR_URL_Pattern_Prefixer'.
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * Class for prefixing URL patterns.
+ *
+ * @since n.e.x.t
+ */
+class PLSR_URL_Pattern_Prefixer {
+
+	/**
+	 * Map of `$context_string => $base_path` pairs.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	private $contexts;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array $contexts Optional. Map of `$context_string => $base_path` pairs. Default is the contexts returned
+	 *                        by the {@see PLSR_URL_Pattern_Prefixer::get_default_contexts()} method.
+	 */
+	public function __construct( array $contexts = array() ) {
+		if ( $contexts ) {
+			$this->contexts = array_map( 'trailingslashit', $contexts );
+		} else {
+			$this->contexts = self::get_default_contexts();
+		}
+	}
+
+	/**
+	 * Prefixes the given URL path pattern with the base path for the given context.
+	 *
+	 * This ensures that these path patterns work correctly on WordPress subdirectory sites, for example in a multisite
+	 * network, or when WordPress itself is installed in a subdirectory of the hostname.
+	 *
+	 * The given URL path pattern is only prefixed if it does not already include the expected prefix.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $path_pattern URL pattern starting with the path segment.
+	 * @param string $context      Optional. Either 'home' (any frontend content) or 'site' (content relative to the
+	 *                             directory that WordPress is installed in). Default 'home'.
+	 * @return string URL pattern, prefixed as necessary.
+	 */
+	public function prefix_path_pattern( string $path_pattern, string $context = 'home' ): string {
+		// If context path does not exist, the context is invalid.
+		if ( ! isset( $this->contexts[ $context ] ) ) {
+			_doing_it_wrong(
+				__FUNCTION__,
+				sprintf(
+					/* translators: %s: context string */
+					esc_html__( 'Invalid context %s.', 'performance-lab' ),
+					esc_html( $context )
+				),
```

Comment:
Nit, I think generally it's better to escape the overall output rather than escape each part. Easier to prevent unescaped content from sneaking in, and easier to review for security.

```suggestion
				esc_html(
					sprintf(
						/* translators: %s: context string */
						__( 'Invalid context %s.', 'performance-lab' ),
						$context
					)
				),
```